harness/determined/_train_context.py

```diff
@@ -196,58 +264,147 @@ class DistributedContext:
 
     def __init__(
         self,
-        env: det.EnvContext,
-        hvd_config: horovod.HorovodContext,
+        rank_info: RankInfo,
         rendezvous_info: det.RendezvousInfo,
+        unique_port_offset: int,
+        force_tcp: bool = False,
     ) -> None:
-        self._env = env
-        self._hvd_config = hvd_config
+        self._info = rank_info
         self._rendezvous_info = rendezvous_info
+        self._unique_port_offset = unique_port_offset
 
-        if self._hvd_config.use:
-            self._is_chief = horovod.hvd.rank() == 0
-        else:
-            self._is_chief = True
+        self._is_chief = self._info.rank == 0
+        self._is_local_chief = self._info.local_rank == 0
 
-        if self._hvd_config.use:
-            # Initialize zmq comms.
-            srv_pub_port = (
-                constants.INTER_TRAIN_PROCESS_COMM_PORT_1 + self._env.det_trial_unique_port_offset
+        if len(self._rendezvous_info.get_addrs()) != self._info.cross_size:
+            raise AssertionError(
+                f"rendezvous_info has {len(self._rendezvous_info.get_addrs())} addresses but "
+                f"rank_info indicates there are {self._info.cross_size} machines"
+            )
+
+        self._init_ipc(force_tcp)
+
+    def _init_ipc(self, force_tcp: bool) -> None:
+        if self._info.size < 2:
+            # No broadcasting necessary.
+            return
+
+        # Why does this work?  Why don't multiple agents in training decide on different ports,
+        # causing them to be unable to communicate?
+        #
+        # Answer: unique port offsets are set on by the determined-master on a per-container basis.
+        # Multi-agent dtrain only allows for one-container-per-node, in which case
+        # unique_port_offset is always 0, and every container in training will agree on the same
+        # port.  Single-agent dtrain always within a single container, so the unique_port_offset
+        # may not be zero but all workers (in that same container) will still agree.
```

Comment:
```suggestion
        # unique_port_offset is used for avoiding collision of ports when using distributed training.
        # It should work even if containers use host network. unique_port_offset is set by the master
        # on a per-container basis. With multi-agent dtrain, we run only one container on each node, 
        # in which case unique_port_offset is 0 for all the containers. With single-agent dtrain, we 
        # run only one container, unique_port_offset might not be zero.
```

Reply:
Okay. I am okay with removing this QA. I was simply suggesting to use declarative sentences rather than QA.